py/formatfloat: Improve accuracy of float formatting code.

yoctopuce opened the pull request:
### Summary

Following discussions in [PR #16666](https://github.com/micropython/micropython/pull/16666), this pull request updates the float formatting code to reduce the `repr` reversibility error, i.e. the percentage of valid floating point numbers that do not parse back to the same number when formatted by `repr`.

The baseline before this commit is an error rate of ~46%, when using double-precision floats.

This new code is available in two flavors, based on a preprocessor definition:
- In the simplest version, it reduces the error down to ~40%, using an integer representation of the decimal mantissa rather than working on floats. It is also slightly faster, and improves the rounding in some conditions.
- In the most complete version, it reduces the error down to ~5%. This extra code works by iterative refinement, and makes the code slightly slower than CPython when tested on ports/unix.

### Testing

The new formatting code was tested for reversibility using the code provided by Damien in [PR #16666](https://github.com/micropython/micropython/pull/16666)
A variant using formats `{:.7g}`, `{:.8g}` and `{:.9g}` was used for single-precision testing.

Compatibility with CPython on the various float formats was tested by comparing the output using the following code:
```py
for mant in [34567, 76543]:
    for exp in range(-16, 16):
        print("Next number: %de%d" % (mant, exp))
        num = mant * (10.0**exp)
        for mode in ['e', 'f', 'g']:
            maxprec = 16
            # MicroPython has a length limit in objfloat.c
            if mode == 'f' and 6 + exp + maxprec > 31:
                maxprec = 31 - 6 - exp
            for prec in range(1, maxprec):
                fmt = "%." + str(prec) + mode
                print("%5s: " % fmt, fmt % num)
```
The integration tests have also found some corner cases in the new code which have been fixed.
For single-precision floats, some test cases had to be adapted:
- `float_format_ints` is tapping into an ill-defined partial digit of the mantissa (the 10th), which is not available in single-precision floats with the new code due to integer limitations. So the display range has been updated accordingly.
- similarly, `float_struct_e` uses a 15-digit representation which is meaningless on single-precision floats. A separate version for double-precision has been made instead
- in `float_format_ints`, there is one test case specific to single-precision floats which verifies that the largest possible mantissa value `16777215` can be used to store that exact number and retrieve it as-is. Unfortunately the rounding in the simplest version of the new algorithm makes it display as a slightly different number. This would cause the CI test to fail on single-precision floats when the improved algorithm is not enabled.

### Trade-offs and Alternatives

It is unclear at that point if the simplest version of this improvement is worth the change: 
- going from 46% error to 40% error in double precision is not a big improvement.
- there is no improvement for single-precision
- the new code is only marginally faster

The full version of the enhancement makes much more difference in terms of precision, both for double-precision and single-precision floats, but it causes about 20% overhead on conversion time, and makes the code a bit bigger

Looking forward to reading your feedback...

Edit 1: See https://github.com/micropython/micropython/pull/17444#issuecomment-2979968660 for updates on accuracy results
Edit 2: Updated values in https://github.com/micropython/micropython/pull/17444#issuecomment-2987116217

yoctopuce committed:
py/formatfloat: Improve accuracy of float formatting code.

Following discussions in PR #16666, this commit updates the float
formatting code to improve the `repr` reversibility, i.e. the percentage of
valid floating point numbers that do parse back to the same number when
formatted by `repr` (in CPython it's 100%).

This new code offers a choice of 3 float conversion methods, depending on
the desired tradeoff between code size and conversion precision:

- BASIC method is the smallest code footprint

- APPROX method uses an iterative method to approximate the exact
  representation, which is a bit slower but but does not have a big impact
  on code size.  It provides `repr` reversibility on >99.8% of the cases in
  double precision, and on >98.5% in single precision (except with REPR_C,
  where reversibility is 100% as the last two bits are not taken into
  account).

- EXACT method uses higher-precision floats during conversion, which
  provides perfect results but has a higher impact on code size.  It is
  faster than APPROX method, and faster than the CPython equivalent
  implementation.  It is however not available on all compilers when using
  FLOAT_IMPL_DOUBLE.

Here is the table comparing the impact of the three conversion methods on
code footprint on PYBV10 (using single-precision floats) and reversibility
rate for both single-precision and double-precision floats.  The table
includes current situation as a baseline for the comparison:

              PYBV10  REPR_C   FLOAT  DOUBLE
    current = 364688   12.9%   27.6%   37.9%
    basic   = 364812   85.6%   60.5%   85.7%
    approx  = 365080  100.0%   98.5%   99.8%
    exact   = 366408  100.0%  100.0%  100.0%

Signed-off-by: Yoctopuce dev <dev@yoctopuce.com>

yoctopuce commented:
Additional results after a few more experiments in double-precision :
- with the simple method, %17g provides the fastest `repr` value with ~40% error 
- with the elaborate method enabled, we can bring `repr` error down to 0.5% as follows:
  1) try first %17g, and stop if a perfect match is found (90% of the cases)
  2) otherwise fallback to %18g (7% of the cases)
  3) otherwise fallback to %19g (remaining cases)
  The overall cost is only 20% more than the fastest `repr` that had 40% error.

dpgeorge commented:
Thanks for this.  I will study it in detail.

I also found the corresponding CPython issue and long discussion about this topic: https://github.com/python/cpython/issues/45921 .  There's a lot of useful information there.  It looks like this is a hard thing to get right.

yoctopuce commented:
> I also found the corresponding CPython issue and long discussion about this topic: [python/cpython#45921](https://github.com/python/cpython/issues/45921) . There's a lot of useful information there. It looks like this is a hard thing to get right.

I read that thread, it is indeed an interesting input. But I am afraid that we will not be able to use much of it due to ressource constraints in MicroPython.

yoctopuce commented:
> * with the elaborate method enabled, we can bring `repr` error down to 0.5% as follows:

Out of curiosity, I have been investigating a bit more...

Half of these remaining errors are caused by `{:g}` falling back to `{:f}` for small negative exponents and therefore loosing significant digits in the mantissa. If we run the test using `{:e}`, the `repr`error goes down to 0.25%. So I am looking at changing the code for `{:f}` to insert leading zeroes without integrating them in the mantissa. This should properly fix the issue and improve the accuracy of both `{:f}` and `{:g}`.

About 20% of the remaining errors are caused by large negative exponents (exp < -255). There is probably something we could improve there as well, but I am not sure it is a big concern for MicroPython if such seldom-used numbers have accuracy problems in `repr`. 

I am also looking at cleaning a bit the code, to reduce code footprint. While doing this, one things that I noticed is that the function includes lots of checks on `buf_size` and `buf_remaining` at various places, but MicroPython always uses a 32 bytes buffer (or 16 bytes buffer for single-precision floats). Don't you think it would make sense to document this function as requiring a 32 bytes output buffer, adding an `assert(buf_size >= 32)` at the beginning and simplifying these checks ?

dpgeorge commented:
> Don't you think it would make sense to document this function as requiring a 32 bytes output buffer, adding an `assert(buf_size >= 32)` at the beginning and simplifying these checks ?

Yes, that makes a lot of sense.

dpgeorge commented:
Related: #6024 has been around for a while and attempt to improve the other side of this, ie parsing of floats, using higher precision ints.

yoctopuce commented:
The unix / nanbox variant crash at the end of the accuracy test, due to the randomly generated float number that create invalid nanbox objects.

I tried a quick fix by excluding  this specific test based on the platform name, but obviously `nanbox` is not a platform name so that does not work. Any idea how I could detect nanbox to skip this test ?

jepler commented on py/mpprint.h:
```diff
@@ -37,7 +37,9 @@
-#define PF_FLAG_SEP_POS           (9) // must be above all the above PF_FLAGs
+#define PF_FLAG_USE_OPTIMAL_PREC  (0x200)
+#define PF_FLAG_ALWAYS_DECIMAL    (0x400)
+#define PF_FLAG_SEP_POS           (16) // must be above all the above PF_FLAGs
```
Over in the PR where the code that added PF_FLAG_SEP_POS we discussed the position of this flag. There is a port with 16-bit `int`s. It was possible to position PF_FLAG_SEP_POS at 9, because there were enough bits in an unsigned 16-bit value to squeeze in `_` and `,` as shifted values without widening the type of the flags argument.

It's my mistake that I didn't comment on the requirement here, nor write a check that would trigger during CI. There is an assertion about the situation in `objstr.c` but because the pic16bit port is not built during CI (non-free toolchain) the assertion is only checked on platforms with 32-bit ints.

yoctopuce replied:
Goot point. I saw your (unrelated) PR on `mpprint.h` yesterday and was wondering about this.
I have just updated the code to move these new flags above the `SEP` bits.
As they are strictly related to the use of floats, which are obviously not enabled in the pic16 port, this should be safe.

yoctopuce commented:
Heads up: I found a trick to get 100% good conversions, significantly faster then CPython, and the code appears to be smaller than my previous version. Working on it...

yoctopuce commented:
I have finally managed to complete cleaning my new implementation. It is a much bigger rewrite than I anticipated originally...

This new code offers a choice of 3 float conversion methods, depending on the desired tradeoff between code size and conversion precision:
- BASIC method is the smallest code footprint: it achieves better results than current method through the use of `mp_float_uint_t` to work on the mantissa, and also by trying multiple precisions to find the best `repr`.
- APPROX method iteratively improves the mantissa, which is a bit slower but but does not have a big impact on code size. It provides `repr` reversibility on >99.70% of the cases.
- EXACT method uses higher-precision floats during conversion,  which provides best results, but has a higher impact on code size. It is faster than APPROX method, and faster than  CPython equivalent implementation.

Here is the table comparing the impact of the three conversion methods on code footprint on PYBV10 (using single-precision
floats) and reversibility rate for both single-precision and double-precision floats. The table includes current situation as a baseline for the comparison:

|               | PYBV10  |  IMPL|  IMPL
|               | code size  |  FLOAT |  DOUBLE
|----------|------------|--------|--------|
|current implementation | 364136 |  85.47% |  37.90% |
|MICROPY_FLOAT_CONV_BASIC| 364188  | 97.78%  | 62.18% |
|MICROPY_FLOAT_CONV_APPROX| 364396 |  99.70% |  99.84% |
|MICROPY_FLOAT_CONV_EXACT| 365608  |100.00% | 100.00% |

Unfortunately, on ARM gcc, the EXACT method is significantly bigger than the APPROX method, so I am not sure this is worth the code size increase. However on platforms that use IMPL_DOUBLE, I guess it might be worth enabling it.

The commit also include two minor fix for nanbox, that were preventing the new CI tests to run properly on that port.

I will look for any CI test failing.

robert-hh commented:
The slowest board in the previous test was the ESP8266. So I ran run-tests.py -t u0 float/*.py. It ran to an end, with the test summary below:

42 tests performed (2417 individual testcases)
34 tests passed
24 tests skipped: float/cmath_dunder.py float/cmath_fun.py float/cmath_fun_special.py float/complex1_intbig.py float/complex1.py float/complex_dunder.py float/complex_reverse_op.py float/complex_special_methods.py float/float2int_doubleprec_intbig.py float/float2int_intbig.py float/float_divmod.py float/float_format_ints_doubleprec.py float/float_format_ints_power10.py float/float_parse_doubleprec.py float/float_struct_e_doubleprec.py float/int_big_float.py float/math_constants_extra.py float/math_domain_special.py float/math_factorial_intbig.py float/math_fun_special.py float/math_isclose.py float/string_format.py float/true_value.py float/types.py
8 tests failed: float/builtin_float_round.py float/float1.py float/float_format_accuracy.py float/float_format_ints.py float/float_format.py float/float_parse.py float/lexer.py float/python36.py

Running the test on a SAMD21, only float/float_format_accuracy.py failed.
Renesas-RA EK-RA6M2 is slow as well, but all tests eithe pass or are skipped.

The tested ESP32, RP2 and NRF boards fail the float_format_accuracy.py test, because it takes too long.
ESP8266 and SAM21 fail this test with:

```
  File "<stdin>", line 19, in pseudo_randouble
AttributeError: 'module' object has no attribute 'crc32'
```

robert-hh commented:
The test  `float_format_accuracy.py` now passes for ESP32, NRF and RP2. The runtime is betwee 4-5 seconds. 
The test `float/float_format_accuracy.py` runs fine to an end with ESP8266 and SAMD21 as well, but the execution time is too long with 15 and 17 seconds. So running it with run-test.py will fail.

Besides that I wonder about the function pseudo_randouble(). As far as I can tell, it fills a list with pseude-random numbers with the first value always being 42, converts that list to a bytes-object and that to a float and returns it. So my question:

a) why do you not use random.getrandbits() for the number? It is available at most ports & boards. It uses a pseudo-rng. Trying to use random.getrandbits leads to spurious fails like this one: `FAILED: repr rate=99.950% max_err=2.720e-07`

b) pseudo_randouble() returns both that random number list ddef and the float value. But ddef is not used. Dropping ddef as return value improves the speed a little bit.

yoctopuce commented:
> The test `float_format_accuracy.py` now passes for ESP32, NRF and RP2. The runtime is betwee 4-5 seconds. The test `float/float_format_accuracy.py` runs fine to an end with ESP8266 and SAMD21 as well, but the execution time is too long with 15 and 17 seconds. So running it with run-test.py will fail.

OK, I will try to reduce the number of samples even more.

> a) why do you not use random.getrandbits() for the number? It is available at most ports & boards. It uses a pseudo-rng. 

I was afraid that `getrandbits()` would not be available everywhere, as it depends on `MICROPY_PY_RANDOM_EXTRA_FUNCS` which is only available with EXTRA features. By using a crc32-based pseudo generator, I was doing about the same, in a way that would be for sure reproductible accross all platforms.

> Trying to use random.getrandbits leads to spurious fails like this one: `FAILED: repr rate=99.950% max_err=2.720e-07`

If the number of samples if too small, the status may not converge soon enough. My guess is that if you let it run for 10'000 samples or more, the stats converge to
- 99.70% for IMPL_FLOAT
- 99.83% for IMPL_DOUBLE
regardless of the random generator.

> b) pseudo_randouble() returns both that random number list ddef and the float value. But ddef is not used. Dropping ddef as return value improves the speed a little bit.

Good point. I used to save ddef in order to be able to analyze number that caused problems, but this is no more needed. I will remove it.

Thank you again for all the time you spend running these test, it really helps a lot !

yoctopuce commented:
> Your solution for pseudo_randouble() generates always the same sequence of numbers with the same number of differences.

Correct, that was the intent. We want to test a variety of floats, but we want the test to be reproductible, to avoid random failures.

> Another observation: If I print f and f2 in the test branch where the match, the majority of numbers print a inf or -inf. 

Oho, good point. As I was developing on `ports/unix` with a fast machine, I did not realize that my random generator was always based on doubles, which means a large quantity of generated numbers are actually dropped when testing on a single-precision port. This is stupid, as it waste precious testing. I will change the generator to generate 32 bit floats on 32 bit machines, rather than generating 64 bit floats and converting them to 32 bits. This should speed up things a lot !

Thank you again, will post a new version tomorrow morning.

robert-hh commented:
Some more results:
- Testing proper single precision floats shows a  higher fail rate: `FAILED: repr rate=99.300% max_err=2.810e-07`
- Using getrandbits() at ESP8266 fails with a strange error after 418 attempts `TypeError: can't convert NoneType to float`.  It seems that sometimes other types that float are returned by pseudo_randsingle() (I made two functions. If I change the test for proper numbers to: `while type(f) is not float or math.isinf(f) or math.isnan(f) or math.frexp(f)[1] <= min_expo:`, then it passes on ESP8266, and the time for 2000 values is below 8 seconds using getrandbits().
- No need to support SAMD21 for the test. SAMD51 can be used for the SAMD port and is much faster (1s instead of 12s).

yoctopuce commented:
> Some more results:
> 
> * Testing proper single precision floats shows a  higher fail rate: `FAILED: repr rate=99.300% max_err=2.810e-07`
> * Using getrandbits() at ESP8266 fails with a strange error after 418 attempts `TypeError: can't convert NoneType to float`.  It seems that sometimes other types that float are returned by pseudo_randsingle() (I made two functions. If I change the test for proper numbers to: `while type(f) is not float or math.isinf(f) or math.isnan(f) or math.frexp(f)[1] <= min_expo:`, then it passes on ESP8266, and the time for 2000 values is below 8 seconds using getrandbits().
> * No need to support SAMD21 for the test. SAMD51 can be used for the SAMD port and is much faster (1s instead of 12s).

Thank you for these additional tests.

I have just pushed a new version of the test code, that fixes , with a fixed version of the floating point generator and should make it faster by avoiding garbage-collecting the list.

I realized that the previous version that was generating doubles, then casting them to single-precision floats was significantly biasing the results for single precision floats, by running a lots of tests on `inf` (due to the wrong test) and on '0.0'. This also changes the stats for the baseline using existing code, which is much worth than initially thought. So although the accuracy obtained by the APPROX method is now significantly lower than initially thought, it also shows that the new code is a big improvement over the existing one.

Here are the updated stats (based on 1 Mio random numbers):

|               | PYBV10  code size  |  IMPL_FLOAT |  IMPL_DOUBLE |
|----------|------------|--------|--------|
|current implementation | 364136 |  27.57% |  37.90% |
|MICROPY_FLOAT_CONV_BASIC| 364188  | 91.01%  | 62.18% |
|MICROPY_FLOAT_CONV_APPROX| 364396 |  98.50% |  99.84% |
|MICROPY_FLOAT_CONV_EXACT| 365608  |100.00% | 100.00% |

Would you mind giving a try to this new test code ?

robert-hh commented:
I will test that during the day. Just a quick compare for the two ends of the range of your recent test accuracy code with the code I used:

MIMXRT 1.4s (your code) vs 0.37s
ESP8266 39s vs 7.5 s

Note that even with your code the ESP8266 needs the changed test, that the value returned by pseudo_randouble() is a float.

yoctopuce commented on tests/float/float_format_accuracy.py:
```diff
@@ -0,0 +1,115 @@
+max_err = 0
+for _ in range(N):
+    f = pseudo_randouble()
+    while math.isinf(f) or math.isnan(f) or math.frexp(f)[1] <= min_expo:
```
Sorry, I did not realize the problem would not be specific to `randbits`. I have now force-pushed a new version reusing many of yours ideas, including using `randbits`. I have limited the number of tests to 1500, so it should run on all platforms I guess.

Would you give it a try ?

robert-hh commented:
There is not much difference between rnd_buf[] being local or global. The local version is slightly faster, since it does not need a symbol lookup. So for code clarity the local version seems better. Interestingly the results for ESP8266 and SAMD21, RP2, STM32 or ESP32 at N=1200 are different. Except for SAMD21, you could go back to N=2000.

SAMD21 at N=1200:
```
1200 values converted
FAILED: repr rate=98.167% max_err=8.675e-08

real	0m7,237s
user	0m0,109s
sys	0m0,038s
```
ESP8266, N=1200:
```
1200 values converted
FAILED: repr rate=98.750% max_err=3.082e-07

real	0m4,653s
user	0m0,114s
sys	0m0,027s
```
ESP8266, N=2000:
```
2000 values converted
FAILED: repr rate=98.900% max_err=3.082e-07

real	0m7,240s
user	0m0,131s
sys	0m0,021s
```
RP2 at N=1200:
```
1200 values converted
FAILED: repr rate=98.167% max_err=8.675e-08

real	0m2,802s
user	0m0,106s
sys	0m0,024s
```
RP2 at N=2000:
```
2000 values converted
float format accuracy OK

real	0m4,436s
user	0m0,126s
sys	0m0,009s
```

robert-hh commented:
On a side note: At single precision the function pseudo_randfloat() is called 2301 times, at double precision it's 2171.

robert-hh commented:
So they all pass fine at N=2000 except for the ESP8266, where the error stays too high. I wonder what's the difference using  the same float code in MP. Obviously the compiler & C-libs are different.

yoctopuce commented:
> > 1. do you prefer that I include the fix in this PR as I did for nanbox, or that I move both to a different PR ?
> 
> Please move to a different PR, this one is getting big :)

@dpgeorge I have submitted a PR with the trivial fix for `math.nan` and hand-crafted floats in `nanbox`. Once it will be integrated into master, I can remove that part of the code from this PR and rebase to the new master. 

I have intentionally kept the REPR_C crafted-floats issue separate, as more time will be needed to find the most efficient implementation. The test code is already available here: https://github.com/micropython/micropython/commit/edf5cab28ab7ebd953cbb2e91b73a98be834bab5

yoctopuce commented:
I have rebased the PR to master and removed nanbox-related changes, now that they have been integrated in master.

dpgeorge commented on py/mpprint.c:
```diff
@@ -353,6 +353,14 @@ int mp_print_float(const mp_print_t *print, mp_float_t f, char fmt, unsigned int
     char *s = buf;
 
+    if ((flags & PF_FLAG_ALWAYS_DECIMAL) && strchr(buf, '.') == NULL && strchr(buf, 'e') == NULL && strchr(buf, 'n') == NULL) {
+        if ((size_t)(len + 2) < sizeof(buf)) {
```
Is it possible to remove this buffer size check by allocating 2 more bytes to `buf`, and then passing `sizeof(buf) - 2` to `mp_format_float()` above, thereby guaranteeing there will always be enough space to add these characters?

Same comment for the `PF_FLAG_ADD_PERCENT` check below, so add 3 bytes in total to `buf` and actually pass `sizeof(buf) - 3` to `mp_format_float()`.

dpgeorge commented on py/mpconfig.h:
```diff
@@ -856,6 +856,23 @@ typedef double mp_float_t;
+#endif
+
+#ifndef MICROPY_FLTCONV_IMPL
+#define MICROPY_FLTCONV_IMPL (MICROPY_FLOAT_DEFAULT_FLTCONV_IMPL)
```
Is this `MICROPY_FLOAT_DEFAULT_FLTCONV_IMPL` needed?  Why not just wrap the above bit of code in `#ifndef MICROPY_FLTCONV_IMPL` and then define `MICROPY_FLTCONV_IMPL` directly?

Also, maybe call it `MICROPY_FLOAT_FORMAT_IMPL`?

yoctopuce commented:
> Definitely `mpy-cross` should use the exact algorithm, because that benefits all targets and the cost (code size and speed) is host side. Could you enable that?

The default settings in `mpconfig.h` should enable `EXACT` mode automatically by default for `mpy-cross` when using `gcc`, because `mpy-cross` is using `IMPL_DOUBLE` and `long double` is available. On Windows, there is no true `long double`, so `APPROX` is the best we can do anyway...

dpgeorge commented on py/parsenum.c:
```diff
@@ -178,27 +180,77 @@ typedef enum {
 #endif
 
+// Helper to compute `num * (10.0 ** dec_exp)`
+mp_large_float_t mp_decimal_exp(mp_large_float_t num, int dec_exp) {
```
This refactoring of `parsenum.c` -- and also it looks like it slightly improves things by doing 5**exp instead of 10**exp -- can you put it in a separate commit?  It would be good to have that change self-contained so that it's easier to understand what is going on.

yoctopuce committed:
py/parsenum: Refactor float parsing code.

This commit extracts from the current float parsing code two functions
which could be reused elsewhere in MicroPython.

The code used to multiply a float x by a power of 10 is also simplified by
applying the binary exponent separately from the power of 5.  This avoids
the risk of overflow in the intermediate stage, before multiplying by x.

Signed-off-by: Yoctopuce dev <dev@yoctopuce.com>

yoctopuce commented on py/mpconfig.h:
```diff
@@ -856,6 +856,21 @@ typedef double mp_float_t;
 #define MICROPY_PY_BUILTINS_COMPLEX (MICROPY_PY_BUILTINS_FLOAT)
 #endif
 
+// Float conversion to/from string implementations
```
Let me know if the added text is OK

dpgeorge replied:
Yes, that's good now.

yoctopuce commented on py/misc.h:
```diff
@@ -272,6 +272,57 @@ typedef union _mp_float_union_t {
     mp_float_uint_t i;
 } mp_float_union_t;
 
+static inline int fp_signbit(mp_float_t x) {
```
You definitely have a talent to find where code could be simplified :-)

These definitions were directly inherited _as is_ from the previous version of the format code. But that code was clearly not up-to-date with current usage of floating point tools in MicroPython: it was duplicating many helper functions which are now available from other places. I have now cleaned up most of these, and kept the last ones which might be useful directly in `py/formatfloat.c`

dpgeorge replied:
> These definitions were directly inherited _as is_ from the previous version of the format code.

Ah, I see.



> I have now cleaned up most of these, and kept the last ones which might be useful directly in `py/formatfloat.c`

Very good!

yoctopuce commented:
OK, it looks like we are all set for the next review, the last failing test is a false positive.

dpgeorge commented:
One last thing: it would be good to have a build that uses the BASIC mode.  I think esp8266 is the ideal candidate for that because it has little flash available, not much computing resources, and if we have room to spare on that port it's best to use that room for other things.

Could you enable BASIC on that port, in `ports/esp8266/mpconfigport.h`?

I tested and with APPROX (current situation) that port grows by +380 bytes, with BASIC it grows by +132.

yoctopuce commented:
Quick update: still working on it, I am almost done (but it takes a bit longer as I am out of office these days...)

yoctopuce commented:
I have pushed a new version which fixes most of the issues with `REPR_C`. The new code applies the optimal rounding method by adding mantissa digits progressively to all formatting operations where it could be beneficial, rather than just for `repr`. This provides a much nicer output for all number based on decimal exponents. I have been comparing the output of the following test code with CPython and using all float formats:
```
for mant in [34567, 76543, 999999]:
    for exp in range(-16, 16):
        strnum = "%de%d" % (mant, exp)
        print("Next number: %s" % strnum)
        num = float(strnum)
        for mode in ['e', 'f', 'g']:
            maxprec = 16
            # MicroPython has a length limit in objfloat.c
            if mode == 'f' and 6 + exp + maxprec > 31:
                maxprec = 31 - 6 - exp
            for prec in range(0, maxprec):
                fmt = "%." + str(prec) + mode
                print("%5s: " % fmt, fmt % num)
```
With the latest code, even REPR_C outperforms CPython in providing a better representation without "phantom" digits.

Here is the code size and format reversibility score, including current method as a baseline for comparison:

|           |  PYBV10 code size | REPR_C score |  FLOAT score | DOUBLE score |
|--------|---------|---------|--------|---------|
|current | 364596 |  12.9% |  27.6% |  37.9% |
|basic   | 364712  | 85.6% |   60.5% |  85.7% |
| approx  | 364964 | 100.0% |  98.5% |  99.8% |
| exact  | 366408  | 100.0% | 100.0% |  100.0% |

I think there is still some room for improvement in this code, at least for one REPR_C rounding case, and to reduce the code size. But you can already give a try of the new code on esp8266 with REPR_BASIC.

yoctopuce commented:
Interesting, this failure is not really surprising but I could not reproduce it on a unix build, even when using REPR_C in BASIC or APPROX mode.

I have nevertheless added the tolerance to the missing bits for this test as well, so now it should now be ok.

dpgeorge commented:
esp8266 now works OK in APPROX mode.

In BASIC mode I get:
```
FAILURE micropython/tests/results/float_float_format_ints.py                     
--- micropython/tests/results/float_float_format_ints.py.exp
+++ micropython/tests/results/float_float_format_ints.py.out
@@ -538,6 +538,6 @@
 23456 x 10^4 with format {:.9e} gives 2.345600000e+08
 23456 x 10^4 with format {:.9f} gives 234560000.000000000
 23456 x 10^4 with format {:.9g} gives 234560000
-16777215.000000
+16777214.400000
 2147483520.000000
 1.000000e+38

FAILURE micropython/tests/results/float_float_format_accuracy.py                      
--- micropython/tests/results/float_float_format_accuracy.py.exp
+++ micropython/tests/results/float_float_format_accuracy.py.out
@@ -1,2 +1,2 @@
 1200 values converted
-float format accuracy OK
+FAILED: repr rate=83.000% max_err=5.737e-07
```
I couldn't reproduce those errors on unix longlong variant in BASIC mode... so maybe that's due to the esp8266's math library and/or soft float math functions?

It would be nice to have the test suite pass in BASIC mode, although maybe it's not fully necessary.

We need to also decide if esp8266 uses BASIC or APPROX.  I was pushing for BASIC (see above), but would be fine if it used APPROX (and then the tests already pass).  BASIC costs +120, APPROX costs +468 (compared to master).  What's your opinion there?

yoctopuce commented:
> I couldn't reproduce those errors on unix longlong variant in BASIC mode... so maybe that's due to the esp8266's math library and/or soft float math functions?

Yes, I am quite sure that a small difference in the implementation of `pow()` function was also the reason for the previous failure in APPROX mode.

> It would be nice to have the test suite pass in BASIC mode, although maybe it's not fully necessary.

I think BASIC mode definitely makes sense on ports with limited ressources, not only for code space but also for speed. In particular, when used with REPR_C, the >80% reversibility rate is quite good compared to previous implementation which did <13%. It is quite interesting to see that BASIC mode does not add a single failure to `rge_sm.py` test.

> We need to also decide if esp8266 uses BASIC or APPROX. I was pushing for BASIC (see above), but would be fine if it used APPROX (and then the tests already pass). BASIC costs +120, APPROX costs +468 (compared to master). What's your opinion there?

I agree that it would be good to have at least one standard build using REPR_C + BASIC, to make sure that nobody breaks it. We could use the `unix/longlong` build for that, but given that we found out that the math library of esp8266 is less accurate, maybe it is a good idea to use it as a worst case scenario, also because code size and speed is a concern on that platform.

So I did push a new version which should pass BASIC and REPR_C on esp8266.

dpgeorge commented:
> It is quite interesting to see that BASIC mode does not add a single failure to `rge_sm.py` test.

Maybe because that test uses only 4 digits to print `{:4f}`?


> So I did push a new version which should pass BASIC and REPR_C on esp8266.

Thanks.  I tested on esp8266 in both APPROX and BASIC mode, and they both pass.  Well done!

Now, I think the only thing left is to enable BASIC on esp8266:
```diff
--- a/ports/esp8266/mpconfigport.h
+++ b/ports/esp8266/mpconfigport.h
@@ -106,6 +106,7 @@
 #define MICROPY_PY_OS_URANDOM       (1)
 #define MICROPY_LONGINT_IMPL        (MICROPY_LONGINT_IMPL_MPZ)
 #define MICROPY_FLOAT_IMPL          (MICROPY_FLOAT_IMPL_FLOAT)
+#define MICROPY_FLOAT_FORMAT_IMPL   (MICROPY_FLOAT_FORMAT_IMPL_BASIC)
 #define MICROPY_WARNINGS            (1)
 #define MICROPY_PY_STR_BYTES_CMP_WARN (1)
 #define MICROPY_STREAMS_POSIX_API   (1)
```
I think it makes sense to do that in this PR.

dpgeorge commented:
Now merged!  Thanks for all your effort here!
